polonius-souffle/build.rs

```diff
@@ -0,0 +1,100 @@
+use std::error::Error;
+use std::collections::HashSet;
+use std::path::{Path, PathBuf};
+use std::process::Command;
+
+use glob::glob;
+use which::which;
+
+type Result<T> = ::std::result::Result<T, Box<dyn Error>>;
+
+/// Gets the filename for each "top-level" rulest
+fn get_rulesets() -> Vec<PathBuf> {
+    let result: std::result::Result<Vec<_>, _> = glob("rules/*.dl").unwrap().collect();
+    result.unwrap()
+}
+fn print_rerun_if_changed() {
+    // Rerun if any C++ file changes
+    for file in glob("shims/*").unwrap() {
+        println!("cargo:rerun-if-changed={}", file.unwrap().to_string_lossy());
+    }
+
+    // Rerun if any datalog file changes.
+    for file in glob("rules/**/*.dl").unwrap() {
+        println!("cargo:rerun-if-changed={}", file.unwrap().to_string_lossy());
+    }
+}
+
+fn main() -> Result<()> {
+    print_rerun_if_changed();
+
+    if which("souffle").is_err() {
+        eprintln!("`souffle` not in PATH. Is it installed?");
+        return Err("missing `souffle`".into());
+    }
+
+    let mut cpp_filenames = vec![];
+    let mut known_stems = HashSet::new();
+    for ruleset in get_rulesets() {
+        // Get the common name for this ruleset.
+        let stem = ruleset.file_stem().unwrap().to_str().unwrap();
+
+        // Check that name for duplicates
+        //
+        // Souffle uses a single, global registry for datalog programs, indexed by string.
+        if !known_stems.insert(stem.to_owned()) {
+            eprintln!("Multiple datalog files named `{}`", stem);
+            return Err("Duplicate filenames".into());
+        }
+
+        let cpp_filename = souffle_generate(&ruleset)?;
+        cpp_filenames.push(cpp_filename);
+    }
+
+    for stem in known_stems {
+        // HACK: Souffle adds datalog programs to the registry in the initializer of a global
+        // variable (whose name begins with `__factory_Sf`). Since that global variable is never used
+        // by the Rust program, it is occasionally removed by the linker, its initializer is never
+        // run (!!!), and the program is never registered.
+        //
+        // `-u` marks the symbol as undefined, so that it will not be optimized out.
+        let prog_symbol = format!("__factory_Sf_{}_instance", stem);
+        println!("cargo:rustc-link-arg=-u{}", prog_symbol);
+    }
+
+    let mut cc = cxx_build::bridge("src/ffi.rs");
+
+    for file in cpp_filenames {
+        cc.file(&file);
+    }
+
+    cc.cpp(true)
+        .define("__EMBEDDED_SOUFFLE__", None)
+        .include("./shims")
+        .flag("-std=c++17")
+        .flag("-Wno-unused-parameter")
+        .flag("-Wno-parentheses")
+        .try_compile("souffle")?;
```

Comment:
FYI, unless there is an implicit `-O2` flag added here by `cxx_build` the performance of Soufflé will take quite a hit (I would assume at least 10x, maybe more)

Reply:
[Let Me Rustdoc That For You](https://docs.rs/cc/1.0.10/cc/struct.Build.html#method.opt_level) :smile:

Reply:
> FYI, unless there is an implicit `-O2` or `-O3` flag added here by `cxx_build` the performance of Soufflé will take quite a hit (I would assume at least 10x, maybe more)

It's about 5x from my experiments.

Reply:
> @ecstatic-morse: Let Me Rustdoc That For You 

Yeah, this was what I expected :+1:

Also note that it is possible to compile Soufflé with OpenMP and to specify [SIPS](https://souffle-lang.github.io/tuning#sideways-information-passing-strategy) if performance ever becomes a concern